### PR TITLE
Fixes #20021 - Optionally generate CA public cert with other CAs certs

### PR DESCRIPTION
**Pull request by sean797**
This simple adds `--other-ca-certs` for use with `--gen-ca` for appending other certs to the CA cert generated.

Reason behind this is so I can create a katello-default-ca cert that includes its own CA as well as the others default CA's in a Katello cluster. This is then distributed to Smart Proxies and allows the Proxies to trust all Katello servers, not just the one `foreman-proxy-certs-generate` was ran on.

I will make the relevant changes to puppet-certs once this is merged.

### CODE REVIEW

File: katello_certs_tools/katello_ssl_tool.py
Comment (ekohl) [inline katello_certs_tools/katello_ssl_tool.py]: Since you're using the contextmanager to open the file it will autoclose and this is redundant.

File: katello_certs_tools/katello_ssl_tool.py
Comment (ekohl) [inline katello_certs_tools/katello_ssl_tool.py]: Same here.

**Review by ekohl — CHANGES_REQUESTED** [review]
I wonder if there should be some handling for IO errors like the file does not exist.

**Comment by sean797** [comment]
Thanks for review @ekohl. I think the traceback is fairly self explanatory & most people don't use this tool directly but instead use it via puppet-certs, in which case we can do https://github.com/Katello/puppet-certs/blob/master/manifests/init.pp#L129-L133 to validate the users input.

File: katello_certs_tools/katello_ssl_tool.py
Comment (ekohl) [inline katello_certs_tools/katello_ssl_tool.py]: This loop can also be replaced by ```outfile.writelines(infile)```.

**Review by ekohl — APPROVED** [review]
That's ok then.

File: katello_certs_tools/katello_ssl_tool.py
Comment (sean797) [reply katello_certs_tools/katello_ssl_tool.py]: Yes, lets do that instead

**Review by ekohl — APPROVED** [review]

**Comment by sean797** [comment]
I was just thinking about this a bit more, I don't think this will ever work as it stands every time you currently add a new CA cert from another machine, the machines you are running on CA certificate changes.

e.g.
Node1 has default-ca-A & node2 has default-ca-B
You update node1 to include default-ca-B, but in that process a new node1's default-ca get updated to default-ca-C. Now node1 has default-ca-B, default-ca-C. node2 has default-ca-B. You update node2 to include default-ca-C, but in that process a new node2's default-ca get updated to default-ca-D.

You are then left with the following Public CA certificates. Node1 has default-ca-B, default-ca-C. Node2 has default-ca-C, default-ca-D

Please do not merge, I will update this.

**Comment by sean797** [comment]
Just pushed something that fixes my previous comment. Solution is to add `--other-ca-certs` with `--gen-ca` & `--rpm-only`. Now when called with `--rpm-only` we check if the cert were going to package already contains the ones in `--other-ca-certs`, if not we add them in.

**Comment by sean797** [comment]
Ping can someone take a look at this please, only doing this as I guess most people don't often check this repo :-)
@ekohl @ehelms @jlsherrill others ;)

**Comment by ehelms** [comment]
@sean797 I can put it on my list, I try my hardest to ignore this repo so that it doesn't get changes put into it ;)

How is this different than just creating a bundle of CA certs?

**Comment by sean797** [comment]
@ehelms thanks! 

I don't think it is too different, it's essentially creating a CA bundle from supplied certs plus it's own.

**Comment by sean797** [comment]
@ekohl @ehelms (and any others) could I convince you guys to merge this? I don't know if this is the way we want to solve issues with multiple nodes and/or split deployments, I suggest we merge this and the other associated PR so to allow us (or at least me 😉) to experiment.

Note this also needs the associated installer PR to make good use of it.  https://github.com/theforeman/puppet-certs/pull/160 which does require a rebase.

**Comment by ekohl** [comment]
Right now I still struggle with understanding the SSL tooling katello uses. Given https://github.com/theforeman/forklift/pull/521 where I create a services box, application server and a proxy, could you point out how we'd get certificates on all 3 machines.

If we go further we could split split pulp into a separate machines, then what? Even further is also splitting qpid into a separate machine? If I understand your design correct we'd now have 4 server VMs (candlepin, qpid, pulp & application) and 4 CAs. Taking it to the ultimate step and making all those 4 VMs HA means we end up with 8 CAs? That feels hard to maintain.

**Comment by sean797** [comment]
@ekohl Obviously that scenario would be sub optimal. I think in that scenario I intend to have only the Candlepin nodes as CAs and everything else trusts them, app, Pulp & qpid nodes don't sign any certs and hence access to a CA. Though it needs testing/experimentation.

EDIT:
I would hope that when we split out app, qpidd, pulp nodes ect..  then we can issue certificates from a the candlepin node to them, similar to what we do with Smart Proxies today.

**Comment by beav** [comment]
Any issue with merging? I see it has :+1: from @ekohl already, and it just adds a new option.

**Comment by bkearney** [comment]
@beav Any update on this?

**Comment by beav** [comment]
change has ack and is additive, merging

**Pull request merged**
